### PR TITLE
[Proposal] New storages interface

### PR DESCRIPTION
**Pull request by icewind1991**
Meant to reduce the work of implementing a storage backend and make it easier to use storages by reducing the interface of the storage.

cc @DeepDiver1975 @Xenopathic @PVince81

### CODE REVIEW

**Comment by PVince81** [comment]
CC @labkode

**Comment by icewind1991** [comment]
reduced it some more

**Comment by icewind1991** [comment]
Note that this is not the api most apps will use to access the filesystem, only the one implenented by backends

**Comment by icewind1991** [comment]
And for some methods we can provide common (propably less efficient) polyfills (copy, rename) so a backend wouldn't even need to implement all of it itself

**Comment by labkode** [comment]
@icewind1991  The idea is to to as much as the backend storage can do and avoid operations on a layer on top of the storage like it is now with versions, trashbin, chunked uploads ...

So rename, copy should be implemented in the backend storage plug-in

**Comment by icewind1991** [comment]
@labkode from our side we would like to keep things like versioning, trash, etc seperated as much as possible from the storage backends (while still allowing the flexibility for a use case like yours)

One idea we had was to add something like `getVersionHandler` to the storage api to allow storages to overwrite version, sharing, trash, etc behaviour while defaulting to our own apps for it.

**Comment by icewind1991** [comment]
versions and trash could also be done transparently on the storage level (either have the storage implement it directly or use a storage wrapper), sharing and chunked uploads are more complex since they are not restricted to a single storage

**Comment by labkode** [comment]
@icewind1991 I agree for the versions, trashbin, sharing ... that is exaclty what I discussed with @DeepDiver1975 , @felixboehm ,@butonic and @PVince81 . But don't think that a getRenameHandler or getCopyHandler to provide default poly fills would make the storage interface cleaner.

**Comment by Xenopathic** [comment]
There's an important distinction between things that apply to the storage, and things that apply to the mount of that storage. Things like file action behaviour (e.g. how files are copied, how mimetypes are retrieved etc) is storage behaviour, while things like sharing of files, or encryption, are mount behaviours.

A storage implementation should only be able to control the storage behaviours. Mount behaviours are controlled with the mount options that are passed into the mount manager. Currently they can control encryption, versioning and file update checking, but could be extended to including sharing and any other features as required.

**Comment by icewind1991** [comment]
I was also thinking about moving the cache/index/catalog/whatever from the storage to the mount point, afaik all the cases where an app needs to modify the behavior of those the app has the option to controll the mountpoints for it

**Comment by Xenopathic** [comment]
@icewind1991 Actually, that would make a lot of sense. The storage impl shouldn't care how its files are being indexed, it just needs to care that when it is asked for a file, it can get that file.

**Comment by icewind1991** [comment]
Added interfaces for trash and version capable storage, storage implementations can implement those to denote that they support versioning/trash

**Comment by icewind1991** [comment]
@labkode what are the requirements for the cache/catalog/index from your side

(are you still at the conf?)

File: lib/public/files/storage/idirectoryentry.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCP\Files\Storage;
+
+interface IDirectoryEntry {
+	const META_MIMETYPE = 'mimetype';
```
Comment (labkode) [inline lib/public/files/storage/idirectoryentry.php]: I guess this is the metadata that is returned for every resource (folder or file).
A getFileID() is needed

File: lib/public/files/storage/istorage.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCP\Files\Storage;
+
+interface IStorage {
+	/**
```
Comment (labkode) [inline lib/public/files/storage/istorage.php]: The methods that deals with metadata (has, listDirectory) should be putted in a interface INamespace/IFileCatalog and the storage interface would have a getNamespaceHandler().

**Comment by labkode** [comment]
@icewind1991 

The requirements every file catalog must support are three:

- Provide a hierarchical structure of the resources in the storage
- Propagate the mtime/Etag to the top level
- Provide unique IDs for every resource in the storage

If you have the opportunity to get @DeepDiver1975 in the conf it will worth a lot, because I already discussed this with him and we agreed to some class/interfaces design

**Comment by icewind1991** [comment]
> The requirements every file catalog must support are three:
>
> -   Provide a hierarchical structure of the resources in the storage
-   Propagate the mtime/Etag to the top level
-   Provide unique IDs for every resource in the storage

Ok, that matches the main requirements we have for the cache/catalog

**Comment by labkode** [comment]
@icewind1991  Yes. Indeed I took those requirements from the oc_filecache ;)

**Comment by Xenopathic** [comment]
I think the easiest thing to do is have two interfaces, `IStorage` which implements the storage-level stuff (get file, put file, move file, copy file, delete file etc) and `IFileCatalog` that implements etag propogation, file ID retrieval, mime type storage etc. It may so happen that a single class implements both, for really fancy storage implementations, but for 'dumb' implementations we can have a separate file catalog implementation.

Thoughts?

**Comment by icewind1991** [comment]
I dont think we need anything more then some way (maybe an interface) for a storage to tell core that it takes care of it's own index/catalog, for any other storage we can extend it using wrapper to update our filecache

**Comment by labkode** [comment]
@Xenopathic I think is okay to have these responsibilities separated and the fact of having two interfaces is very clear. But the storage interface could have a method getFileCatalog() that returns a IFileCatalog implementation.

**Comment by icewind1991** [comment]
When moving all meta data operations to an own interface (removing getMeta, listDirectory from the current proposed storage interface), we run into problem with our "default" use case where the index needs to query meta data from the storage implementation

**Comment by icewind1991** [comment]
For our default case we need getMeta/getDirectory type functions both on the storage and the index

**Comment by Xenopathic** [comment]
Right, I had a think about this, and there's a couple of scenarios that we should aim to seamlessly support:

1. Storage does everything, like the CERN storage backend. All metadata queries, etag queries, mtime etc should just go directly to the storage.
2. Mix of storage and DB file catalogue (our current system). Files stored in the storage, external or not, but all metadata is stored in the file catalogue. The storage can provide some of the data, e.g. mtime, but it gets stored in the catalogue for faster access.
3. File catalogue + object store. The object store could be a standard filesystem too, just that we store files as blobs of data with hashes for names, rather than with their real filenames. This can avoid a lot of the Unicode problems we have on other OSes. All metadata is stored exclusively in the file catalogue, and the object store is only for storing the raw data.

The trick is that we need to accommodate all of those usecases with a single, flexible API...

**Comment by labkode** [comment]
@Xenopathic That´s  definitely the way to go. 

I´ll add also that the API should not be limited to those three uses cases, it has to be flexible enough for custom storage plugins, for example, to use the content hash as the object ID.

Related to the Storage interface, do you plan to include sharing functionality inside the storage or as an external interface detached from the Storage interface?

**Comment by Xenopathic** [comment]
@labkode Would the [Flysystem API](http://flysystem.thephpleague.com/api/) provide all the functionality you need? Or, at least, is it extensible enough to implement the features you need within it's framework? It supports the concept of adapters for the storage backends, and plugins to add new behaviour into the filesystem layer.

**Comment by labkode** [comment]
@Xenopathic 

The features we want, like a Sharing interface, a Versions interface of a Trashbin interface are not going to be solved using a extendable filesystem implementations, for example, using the FlyFileSystem plugin or adapters mechanisms.

The problems will be solved when the upstream code assumes that every storage has a plugin for Sharing, Versions or Trashbin and the Version, Trashbin and Sharing APIs query the common storage interface.

The use of adapters for expanding EOS/MyCustom storage to have sharing or versions or whatever we need is useless if the upstream code (the internal design) doesn't take advantage of it, and we end up in the same situation we are today but using the Fly library instead OC's.

If the goal of using the Fly library is to create a minimal storage interface, then it fails.

There are derivated functions like `getMime()`, `getSize()`, `getOtherThingIWant()` that can be obtained from `getMetadata()` function to avoid the pollution of the storage interface.

Also, reads and puts should always be stream based. Read the full contents of a file or put the full contents of a file is: 1) Bad design 2) Derivated from the stream based apporach.

`write()` and `update()` have to converge into one put() and is the storage that deals with the semantics. This is more efficient and avoid extra ops in upper layers.

Delete operations can be englobed into one deleteResource() function. At the end is just a matter to decide which part manages what is going to be deleted: the upstream code calling `getMetadata()` first or the storage implementation.
But giving the option to delete to the storage is better because some fs doesn't care about of what you are deleting (rm -r ...) and you avoid extra ops in upper layers.

Other thing to discuss here is, and what is very important,
Are you thinking on using the Fly interface as common interface to weird storages like EOS or as a core replacement for OC's multiple Storage interfaces ?.

I like the latest. 

The big advantage I see with Fly interface over OC's is that is not local filesystem based (no concept of open files) and there are quite a good amount of implementations to reduce development effort.

Replacing actual OC Storage interfaces with Fly is a good step to going into it, and it's the first requirement to achieve the versions,trashbin and sharing efficient and scalable model.

I've seen that @icewind1991  did some adaptation of Fly to OC here: https://github.com/owncloud/core/blob/master/lib/private/files/storage/flysystem.php

Another important goal, and the one that impacts more the scalability problem is the way OC presents a global  filesystem view to the end user.
There is one root mount point where all the storages are plugged-in. This doesn't scale.

This root/global view have to be just a storage implementation that acts like a bridge to other storages.

Example:  I am a user with three storage defined:

`local://docs`
`swift:///docs`
`eos:///docs`

The I can acces this storages from the end user point of view direclty.

If a want the Oc unscalable-view I plug the `all` pseudo-storage/view and I can access underlyng real storages with OC global view in mind.

`all:///local:///docs`
`all:///swift:///docs`
`all:///eos:///docs`

This way the organization can tune the storages for its convinience.

And even, sharing it would be another pseude-storage/view that could be defined per storage or as a global namespace:

`sharelocal:///other_docs`
`shareall:///other_docs`
 
 Does this sound reasonable ?
 
I am guy that prefer to write some draws on a blackboard before getting into the code, so If this is not clear I'll try my drawing skills and upload the result here.

**Comment by Xenopathic** [comment]
@labkode Wow, that's quite the write-up! Thanks for all the input!

I think oC's storages API has grown quite bloated, and we could really benefit from scaling back to a standard API like Flysystem. I can envision current oC storages (aka \OCP\Files\Storage implementations) being wrapped in a Flysystem adapter, if we decide to go down that route. The real benefit of something like Flysystem is that we can plug in additional 'modules', like sharing, versions, trashbin, and have them use the adapter implementation if it exists, otherwise fallback to using the DB like we do at the moment.

So, my opinion is that we should replace the internal filesystem components with Flysystem, and provide an adapter to wrap existing \OCP\Files\Storages so they continue to work seamlessly. Much of the API is very similar, which helps. What do you think @icewind1991 @DeepDiver1975 @PVince81 ?

As for the second scalability problem: let's focus on one thing at a time :smile:

**Comment by labkode** [comment]
@Xenopathic I mostly agree on your point, but keep in mind the following:

If you provide the non-minimal interface (pure Flysystem without modifications) and implementations start to appear, if then you want to make it minimal (to make it more scalable) you will break backward compatibility with existing storage implementations.

**Scalability, alongside with security and data integrity shouldn't be a feature of the product, but a core part of the development process that deserves special attention and effort.**

**Comment by DeepDiver1975** [comment]
> So, my opinion is that we should replace the internal filesystem components with Flysystem, and provide an adapter to wrap existing \OCP\Files\Storages so they continue to work seamlessly. Much of the API is very similar, which helps. What do you think @icewind1991 @DeepDiver1975 @PVince81 ?

flysystem is just an implementation for some of the storage backends (maybe) - who knows.
We should never base our own work on interfaces of a possible implementation.

We define the interfaces the way we need them - implementations either fit or not.

Not the other way around!

**Comment by Xenopathic** [comment]
@DeepDiver1975 Flysystem doesn't actually provide any implementations for storage backends, it still relies on SDKs and PHP modules for that. It provides a consistent interface to these storages, so it seems to have exactly the same goal as our storages API.

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 Flysystem doesn't actually provide any implementations for storage backends, it still relies on SDKs and PHP modules for that. It provides a consistent interface to these storages, so it seems to have exactly the same goal as our storages API.

I disagree - our interfaces have to work (code AND design wise) without flysystem as well.

flysystem is a potential implementation - nothing more.

our interfaces can for sure be inspired by flysystem - but thats all

**Comment by Xenopathic** [comment]
@DeepDiver1975 I see. Thanks for clarifying

**Comment by icewind1991** [comment]
@labkode does cern's storage implementation handle all the requirements we have on metadata from our side? (etag and mtime propagation specifically)

**Comment by labkode** [comment]
@icewind1991 Yes, CERN's storage backend handles mtime and ETag propagation.

**Comment by icewind1991** [comment]
I think the best way to be flexible enough to handle all the use cases is to separate the various things a storage backend can do (store blobs, handle metadata, etc) into seperate interface and have storage implementations implement whatever they support.

To provide a unified api an adapter is contructed with whatever backend classes needed (local storage+cache or only an eos backend) and handle the various storage operations with the provided implementations.

I've written up my initial idea for the various interfaces/classes: https://gist.github.com/icewind1991/7ae6885c45f60d9e27af

**Comment by PVince81** [comment]
The discussion has been moved here https://github.com/owncloud/core/issues/20394 from the gist (more direct link)

**Pull request closed without merging**